src/classes/emitter.ts

```diff
@@ -1,49 +1,110 @@
 // deno-lint-ignore-file no-explicit-any
+export type EventName = string | symbol;
 export type Listener = (this: Emitter, ...args: any[]) => void;
 
-export default class Emitter {
-	#events: Record<string, Listener[]> = {};
+export type EventMap = Record<EventName, any>;
 
-	on(e: string, func: Listener) {
-		if (typeof this.#events[e] !== "object") {
-			this.#events[e] = [];
+export default interface Emitter<Events extends EventMap = EventMap> {
+	addListener<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	): this;
+	addListener(eventName: EventName, listener: Listener): this;
+	off<T extends keyof Events>(
+		eventName: T,
+		listener: (this: this, ...args: Events[T]) => void,
+	): this;
+	off(eventName: EventName, listener: Listener): this;
+}
+
+type EventsOf<T> = T extends { _emitter: Emitter<infer U> } ? U
+	: T extends Emitter<infer U> ? U
+	: never;
+
+export function once<T extends Emitter, K extends keyof EventsOf<T>>(
+	emitter: T,
+	eventName: K,
+): Promise<EventsOf<T>[K]>;
+export function once(emitter: Emitter, eventName: EventName): Promise<any[]> {
+	return new Promise((res) => {
+		emitter.once(eventName, (...args) => res(args));
+	});
+}
+
+export default class Emitter<Events extends EventMap = EventMap> {
+	#events = new Map<EventName, Listener[]>();
```

Comment:
We can have a better type if using an object.

```ts
class X<Events> {
  #events = {
     [K in keyof Events]?: ((...args: Events[K]) => void)[] = {}
  }
}
```

Reply:
See also https://github.com/mathe42/surrealdb.deno/blob/82dee508405f5662beecaa4624b0dff9bc866afe/src/classes/emitter.ts